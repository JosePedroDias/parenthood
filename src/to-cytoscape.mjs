// https://js.cytoscape.org/

import { pplStringToRelationships, KINDS } from './core.mjs';

export function relationshipsToCytoscape(relationships) {
  const elements = [];

  let nI = 0;
  let eI = 0;

  const knownPeople = new Map(); // person to node id

  function processNode(name, extra = {}) {
    let id = knownPeople.get(name);
    if (!id) {
      id = `n${nI++}`;
      knownPeople.set(name, id);
      elements.push({
        group: 'nodes',
        data: { id, ...extra },
      });
    }
    return id;
  }

  function processEdge(n1id, n2id, extra = {}) {
    const id = `e${eI++}`;
    elements.push({
      group: 'edges',
      data: { id, source: n1id, target: n2id, ...extra },
    });
  }

  for (const { a, b, kind, siblings } of relationships) {
    const aNodeId = processNode(a, { label: a });
    const bNodeId = processNode(b, { label: b });

    processEdge(aNodeId, bNodeId, {
      label: kind,
      color: 'red',
      arrowShape: 'none',
    });

    for (const sib of siblings) {
      const sNodeId = processNode(sib, { label: sib });

      processEdge(aNodeId, sNodeId, {
        label: 'parent_of',
        color: 'green',
        arrowShape: 'vee',
      });
      processEdge(bNodeId, sNodeId, {
        label: 'parent_of',
        color: 'green',
        arrowShape: 'vee',
      });
    }
  }

  return elements;
}

export function processPplString(s) {
  return relationshipsToCytoscape(pplStringToRelationships(s));
}
