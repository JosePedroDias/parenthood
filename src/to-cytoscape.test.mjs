import { processPplString } from './to-cytoscape.mjs';

describe('processPplString', () => {
  it('processes 1 has_been_with relationship w/ 2 siblings', () => {
    expect(
      processPplString(`p1 +/ p2
      s1
      s2`)
    ).toMatchSnapshot();
  });
});
