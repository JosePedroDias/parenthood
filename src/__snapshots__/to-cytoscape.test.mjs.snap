// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processPplString processes 1 has_been_with relationship w/ 2 siblings 1`] = `
Array [
  Object {
    "data": Object {
      "id": "n0",
      "label": "p1",
    },
    "group": "nodes",
  },
  Object {
    "data": Object {
      "id": "n1",
      "label": "p2",
    },
    "group": "nodes",
  },
  Object {
    "data": Object {
      "arrowShape": "none",
      "color": "red",
      "id": "e0",
      "label": "has_been_with",
      "source": "n0",
      "target": "n1",
    },
    "group": "edges",
  },
  Object {
    "data": Object {
      "id": "n2",
      "label": "s1",
    },
    "group": "nodes",
  },
  Object {
    "data": Object {
      "arrowShape": "vee",
      "color": "green",
      "id": "e1",
      "label": "parent_of",
      "source": "n0",
      "target": "n2",
    },
    "group": "edges",
  },
  Object {
    "data": Object {
      "arrowShape": "vee",
      "color": "green",
      "id": "e2",
      "label": "parent_of",
      "source": "n1",
      "target": "n2",
    },
    "group": "edges",
  },
  Object {
    "data": Object {
      "id": "n3",
      "label": "s2",
    },
    "group": "nodes",
  },
  Object {
    "data": Object {
      "arrowShape": "vee",
      "color": "green",
      "id": "e3",
      "label": "parent_of",
      "source": "n0",
      "target": "n3",
    },
    "group": "edges",
  },
  Object {
    "data": Object {
      "arrowShape": "vee",
      "color": "green",
      "id": "e4",
      "label": "parent_of",
      "source": "n1",
      "target": "n3",
    },
    "group": "edges",
  },
]
`;
